arch/posix/core/posix_core.c

```diff
@@ -0,0 +1,272 @@
+ */
+	}
+}
+
+
+/**
+ * Helper function to start a Zephyr thread as a POSIX thread:
+ *  It will block the thread until a __swap() is called for it
+ *
+ * Spawned from posix_core_new_thread() below
+ */
+static void *pc_thread_starter(void *arg)
+{
+	posix_thread_status_t *ptr = (posix_thread_status_t *) arg;
+
+	/*we detach ourselves so nobody needs to join to us*/
+	pthread_detach(pthread_self());
+
+	if (pthread_mutex_lock(&pc_mtx_threads)) {
+/*
+ * We block until all other running threads reach the while loop in
+ * posix_core_wait_until_allowed() and they release the mutex while they wait
+ * for somebody to signal that they are allowed to run
+ */
+		ps_print_error_and_exit(
+			"Posix arch core: error on pthread_mutex_lock()\n");
+	}
+
+/*
+ * The thread would try to execute immediately, so we need to block it until
+ * allowed
+ */
+	pc_wait_until_allowed();
+
+	pc_new_thread_pre_start();
+
+	/*Once allowed, we jump to the actual thread entry point:*/
+	/*TODO include "nano_internal.h"*/
+	extern void _thread_entry(void (*entry)(void *, void *, void *),
+				  void *p1, void *p2, void *p3);
+
+	_thread_entry(ptr->entry_point, ptr->arg1, ptr->arg2, ptr->arg3);
+
+/*
+ * we only reach this point if the thread actually returns which should not
+ * happen
+ */
+	if (POSIX_ARCH_DEBUG_PRINTS) {
+		ps_print_trace(
+			"Posix arch core: Thread %lu ended\n",
+			pthread_self());
+	}
+
+	return NULL;
+}
+
+/**
+ * Called from _new_thread(),
+ * Create a new POSIX thread for the new Zephyr thread.
+ * _new_thread() picks from the kernel structures what it is that we need to
+ * call with what parameters
+ */
+void pc_new_thread(posix_thread_status_t *ptr)
+{
+
+	if (pthread_create(&(ptr->thread_id), NULL,
+			pc_thread_starter, (void *)ptr)) {
+
+		ps_print_error_and_exit(
+				"Posix arch core: error on pthread_create\n");
+	}
+
+	if (POSIX_ARCH_DEBUG_PRINTS) {
+		ps_print_trace(
+			"Posix arch core: Thread %lu created\n",
+			ptr->thread_id);
+	}
+
+}
+
+/**
+ * Called from _IntLibInit()
+ * prepare whatever needs to be prepared to be able to start threads
+ */
+void pc_init_multithreading(void)
+{
+/*
+ * Let's ensure none of the created threads starts until the first thread (the
+ * one running _PrepC) is switched out
+ *
+ * Note that this is not strictly necessary, as all created threads would
+ * block on the posix_core_currently_running_process condition before
+ * actually doing anything assuming that thread id "0" is an invalid pthread_t.
+ *
+ * But doing this, means we do not need to assume there is any invalid system
+ * thread number
+ */
+	if (pthread_mutex_lock(&pc_mtx_threads)) {
+		ps_print_error_and_exit(
+			"Posix arch core: error on pthread_mutex_lock()\n");
+	}
+}
+
+/*
+ * NOTE:
+ * This will work with or without CONFIG_ARCH_HAS_CUSTOM_SWAP_TO_MAIN defined
+ */
+
+/*
+ * NOTE:
+ * we may want to define CONFIG_ARCH_HAS_THREAD_ABORT
+ * and implement a k_thread_abort() to kill threads which are finished.
```

Comment:
I think we might need to something.
What happens if a thread aborts, and that thread object is then re-used?

Reply:
It would be nice to do something yep. (it is in the TODO)
Although how it is now, an aborted thread just waits forever in pc_wait_until_allowed() after it has been swap'ed out.